Background data movement

byrnedj opened the pull request:
This is the second part of: https://github.com/intel/CacheLib/pull/2

igchor reviewed (commented):
> Reviewed 5 of 24 files at r1.
> *[Reviewable](https://reviewable.io/reviews/intel/cachelib/20)* status: 5 of 24 files reviewed, 3 unresolved discussions (waiting on @byrnedj and @vinser52)

___
*[`cachelib/allocator/BackgroundMover.h` line 68 at r1](https://reviewable.io/reviews/intel/cachelib/20#-NCRVb4N25aauNAL4VcC:-NCRVb4OCXMUoXaHD3R-:bzcsjwr) ([raw file](https://github.com/intel/cachelib/blob/a47446594fccd7cd7f5c9f341f628dff2fa5e6db/cachelib/allocator/BackgroundMover.h#L68)):*
> ```objc
> 
>   ~BackgroundMover() override;
>   
> ```

you could try to run clang-format on those new files - it should get rid of some of those unnecessary white chars
___
*[`cachelib/allocator/BackgroundMover-inl.h` line 87 at r1](https://reviewable.io/reviews/intel/cachelib/20#-NCRW5M-0Uu-TrKtljpG:-NCRW5M-0Uu-TrKtljpH:b-udn3jr) ([raw file](https://github.com/intel/cachelib/blob/a47446594fccd7cd7f5c9f341f628dff2fa5e6db/cachelib/allocator/BackgroundMover-inl.h#L87)):*
> ```objc
>     }
> 
>     totalBytesMoved.add(batch * mpStats.acStats.at(cid).allocSize);
> ```

shouldn;t we use `moved` instead of `batch` for this calculation?
___
*[`cachelib/allocator/CacheAllocatorConfig.h` line 629 at r1](https://reviewable.io/reviews/intel/cachelib/20#-NCRY9Dr2ppoYHLukUG4:-NCRXlKf6Wz-mq0t3OmF:bjz6ehp) ([raw file](https://github.com/intel/cachelib/blob/a47446594fccd7cd7f5c9f341f628dff2fa5e6db/cachelib/allocator/CacheAllocatorConfig.h#L629)):*
> ```objc
>   
>   
>   double minAcAllocationWatermark{0.0};
> ```

Rest of the options (sizeThresholdPolicy, etc.) will be in a separate PR, ritght? I think you might want to remove mention of sizeThresholdPolicy from the .md file for now.


<!-- Sent from Reviewable.io -->

igchor reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> perhaps we could just extend the test a little bit

> *[Reviewable](https://reviewable.io/reviews/intel/cachelib/20#-:-NCRZ0md9u9MefyfAv61:blxcjxw)* status: 5 of 24 files reviewed, 3 unresolved discussions (waiting on @byrnedj and @vinser52)

___
*[`cachelib/allocator/tests/AllocatorMemoryTiersTest.h` line 112 at r1](https://reviewable.io/reviews/intel/cachelib/20#-NCRYd4QDan9-0YeLgLT:-NCRYd4QDan9-0YeLgLU:biwv4g2) ([raw file](https://github.com/intel/cachelib/blob/a47446594fccd7cd7f5c9f341f628dff2fa5e6db/cachelib/allocator/tests/AllocatorMemoryTiersTest.h#L112)):*
> ```objc
>     auto perclassPstats = allocator->getBackgroundMoverClassStats(MoverDir::Promote);
> 
>     EXPECT_GT(1, stats.evictionStats.numMovedItems);
> ```

Can we add a more restrictive test? Since we know the values of watermarks we could test if specified percent of memory is free in the first tier after BG workers do their work.


<!-- Sent from Reviewable.io -->

vinser52 reviewed (changes requested):
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/intel/cachelib/20)* status: 5 of 24 files reviewed, 7 unresolved discussions (waiting on @byrnedj, @igchor, and @vinser52)

___
*[`MultiTierDataMovement.md` line 15 at r1](https://reviewable.io/reviews/intel/cachelib/20#-NCRgZd6BPIY5ML4C33O:-NCRgZd72RIKuTqyE9QU:bfeu8io) ([raw file](https://github.com/intel/cachelib/blob/a47446594fccd7cd7f5c9f341f628dff2fa5e6db/MultiTierDataMovement.md#L15)):*
> ```Markdown
> ## Synchronous Eviction and Promotion
> 
> - `disableEvictionToMemory`: Disables eviction to memory (item is always evicted to NVMe or removed
> ```

Do I understand correctly that tiering would not work in that case? Or we are going to use first-free policy for allocations?
___
*[`MultiTierDataMovement.md` line 26 at r1](https://reviewable.io/reviews/intel/cachelib/20#-NCRgvRbDDvWvnH6gyLq:-NCRgvRbDDvWvnH6gyLr:b-kiswd0) ([raw file](https://github.com/intel/cachelib/blob/a47446594fccd7cd7f5c9f341f628dff2fa5e6db/MultiTierDataMovement.md#L26)):*
> ```Markdown
> - `backgroundEvictorIntervalMilSec`: The interval that this thread runs for - by default
> the background evictor threads will wake up every 10 ms to scan the AllocationClasses. Also,
> the background evictor thead will be woken up everytime there is a failed allocation (from
> ```

misprint: thead - > thread
___
*[`MultiTierDataMovement.md` line 33 at r1](https://reviewable.io/reviews/intel/cachelib/20#-NCRhNYhD8Dclu9Y7G8H:-NCRhNYi00C7Tov7MIoS:b-kx9jze) ([raw file](https://github.com/intel/cachelib/blob/a47446594fccd7cd7f5c9f341f628dff2fa5e6db/MultiTierDataMovement.md#L33)):*
> ```Markdown
> - `evictorThreads`: The number of background evictors to run - each thread is a assigned
> a set of AllocationClasses to scan and evict objects from. Currently, each thread gets
> an equal number of classes to scan - but as object size distribution may be unequal - future
> ```

I am not sure that we should write about future. We can change this description when balancing is implemented.
___
*[`MultiTierDataMovement.md` line 62 at r1](https://reviewable.io/reviews/intel/cachelib/20#-NCRhpGF-c87UQM4SMTP:-NCRhpGF-c87UQM4SMTQ:b-oklsxp) ([raw file](https://github.com/intel/cachelib/blob/a47446594fccd7cd7f5c9f341f628dff2fa5e6db/MultiTierDataMovement.md#L62)):*
> ```Markdown
> ## Background Promoters
> 
> The background promotes scan each class to see if there are objects to move to a lower
> ```

Misprint? promotes -> promoters


<!-- Sent from Reviewable.io -->

byrnedj commented:
> > ```gfm
> > ## Synchronous Eviction and Promotion
> > 
> > - `disableEvictionToMemory`: Disables eviction to memory (item is always evicted to NVMe or removed
> > ```
> 
> Do I understand correctly that tiering would not work in that case? Or we are going to use first-free policy for allocations?

Yes. Tiering does not work if this option is on - we could implement first free or some other allocation policy if this is disabled - see https://github.com/intel/CacheLib/pull/10. I removed documentation for admission policies from this version and will add it to https://github.com/intel/CacheLib/pull/10. 

> _[`MultiTierDataMovement.md` line 33 at r1](https://reviewable.io/reviews/intel/cachelib/20#-NCRhNYhD8Dclu9Y7G8H:-NCRhNYi00C7Tov7MIoS:b-kx9jze) ([raw file](https://github.com/intel/cachelib/blob/a47446594fccd7cd7f5c9f341f628dff2fa5e6db/MultiTierDataMovement.md#L33)):_
> 
> > ```gfm
> > - `evictorThreads`: The number of background evictors to run - each thread is a assigned
> > a set of AllocationClasses to scan and evict objects from. Currently, each thread gets
> > an equal number of classes to scan - but as object size distribution may be unequal - future
> > ```
> 
> I am not sure that we should write about future. We can change this description when balancing is implemented.

The implementation is complete - I can add it to this PR or wait until this one has been merged to develop. 

>Can we add a more restrictive test? Since we know the values of watermarks we could test if specified percent of memory is free in the first tier after BG workers do their work.

Done. And reformatted code.

igchor reviewed (commented):
> Reviewed 1 of 24 files at r1, 1 of 10 files at r2, 1 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/intel/cachelib/20)* status: 4 of 24 files reviewed, 5 unresolved discussions (waiting on @byrnedj, @igchor, and @vinser52)

___
*[`cachelib/allocator/tests/AllocatorMemoryTiersTest.h` line 100 at r3](https://reviewable.io/reviews/intel/cachelib/20#-NCf6aOZ2SqVYA8DyFNE:-NCf6aOZ2SqVYA8DyFNF:b2mmpdo) ([raw file](https://github.com/intel/cachelib/blob/8d4f1a96cbe6ddbd1329efc9672a8a3c4b678306/cachelib/allocator/tests/AllocatorMemoryTiersTest.h#L100)):*
> ```objc
> 
>     //wait for bg movers
>     std::this_thread::sleep_for(std::chrono::seconds(1));
> ```

Is there a way to deterministically wait for BG threads completion? Sleeps like this tend to be hard to debug and cause random failures.

Maybe we could just spin until numMovedItems (or freePercent) is what we expect it to be, and then do all the checks?
___
*[`cachelib/cachebench/util/CacheConfig.cpp` line 100 at r3](https://reviewable.io/reviews/intel/cachelib/20#-NCf7Q4DBt1IfkOHXzS5:-NCf7Q4DBt1IfkOHXzS6:beu7ij6) ([raw file](https://github.com/intel/cachelib/blob/a4c4ab370780615027361cd9ce0c4840162d99db/cachelib/cachebench/util/CacheConfig.cpp#L100)):*
> ```cpp
>   if (configJson.count("memoryTiers")) {
>     for (auto& it : configJson["memoryTiers"]) {
>       memoryTierConfigs.push_back(MemoryTierConfig(it).getMemoryTierCacheConfig());
> ```

where is this code now?


<!-- Sent from Reviewable.io -->

byrnedj commented:
> > Reviewed 1 of 24 files at r1, 1 of 10 files at r2, 1 of 3 files at r3.
> > _[Reviewable](https://reviewable.io/reviews/intel/cachelib/20)_ status: 4 of 24 files reviewed, 5 unresolved discussions (waiting on @byrnedj, @igchor, and @vinser52)
> 
> _[`cachelib/allocator/tests/AllocatorMemoryTiersTest.h` line 100 at r3](https://reviewable.io/reviews/intel/cachelib/20#-NCf6aOZ2SqVYA8DyFNE:-NCf6aOZ2SqVYA8DyFNF:b2mmpdo) ([raw file](https://github.com/intel/cachelib/blob/8d4f1a96cbe6ddbd1329efc9672a8a3c4b678306/cachelib/allocator/tests/AllocatorMemoryTiersTest.h#L100)):_
> 
> > ```objc
> >     //wait for bg movers
> >     std::this_thread::sleep_for(std::chrono::seconds(1));
> > ```
> 
> Is there a way to deterministically wait for BG threads completion? Sleeps like this tend to be hard to debug and cause random failures.
> 
> Maybe we could just spin until numMovedItems (or freePercent) is what we expect it to be, and then do all the checks?

Done.

> _[`cachelib/cachebench/util/CacheConfig.cpp` line 100 at r3](https://reviewable.io/reviews/intel/cachelib/20#-NCf7Q4DBt1IfkOHXzS5:-NCf7Q4DBt1IfkOHXzS6:beu7ij6) ([raw file](https://github.com/intel/cachelib/blob/a4c4ab370780615027361cd9ce0c4840162d99db/cachelib/cachebench/util/CacheConfig.cpp#L100)):_
> 
> > ```c++
> >   if (configJson.count("memoryTiers")) {
> >     for (auto& it : configJson["memoryTiers"]) {
> >       memoryTierConfigs.push_back(MemoryTierConfig(it).getMemoryTierCacheConfig());
> > ```
> 
> where is this code now?

Fixed.

igchor reviewed (changes requested):
> Reviewed 11 of 24 files at r1, 5 of 10 files at r2, 1 of 3 files at r3, 5 of 8 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/intel/cachelib/20)* status: 24 of 28 files reviewed, 6 unresolved discussions (waiting on @byrnedj, @igchor, and @vinser52)

___
*[`cachelib/allocator/BackgroundMover.h` line 20 at r4](https://reviewable.io/reviews/intel/cachelib/20#-NDe6Z4w3QkpUs7XjTX9:-NDe6Z4w3QkpUs7XjTXA:b-75ywm7) ([raw file](https://github.com/intel/cachelib/blob/b31c5adcb19a9d315a4a818a8238f6eec0c1d3c8/cachelib/allocator/BackgroundMover.h#L20)):*
> ```objc
> 
> #include <folly/concurrency/UnboundedQueue.h>
> #include <gtest/gtest_prod.h>
> ```

why is this needed?
___
*[`cachelib/allocator/MMTinyLFU-inl.h` line 235 at r4](https://reviewable.io/reviews/intel/cachelib/20#-NDe6yPvDyVSutdrXzkv:-NDe6yPvDyVSutdrXzkw:b-3u8qk4) ([raw file](https://github.com/intel/cachelib/blob/b31c5adcb19a9d315a4a818a8238f6eec0c1d3c8/cachelib/allocator/MMTinyLFU-inl.h#L235)):*
> ```objc
> void
> MMTinyLFU::Container<T, HookPtr>::withPromotionIterator(F&& fun) {
>   throw std::runtime_error("Not supported");
> ```

Can you maybe open some issue to implement this in future?
___
*[`cachelib/external/fbthrift` line 1 at r4](https://reviewable.io/reviews/intel/cachelib/20#-NDe6Fum1zrMHycpmBpA:-NDe6Fum1zrMHycpmBpB:b-r7568g) ([raw file](https://github.com/intel/cachelib/blob/a4c4ab370780615027361cd9ce0c4840162d99db/cachelib/external/fbthrift#L1)):*
> ```
> Subproject commit 74f3a8fb00e3963d1e6f2361fce722688d500417
> ```

Can you revert updating the dependencies? It will cause conflicts later


<!-- Sent from Reviewable.io -->

byrnedj commented:
> > Reviewed 11 of 24 files at r1, 5 of 10 files at r2, 1 of 3 files at r3, 5 of 8 files at r4, all commit messages.
> > _[Reviewable](https://reviewable.io/reviews/intel/cachelib/20)_ status: 24 of 28 files reviewed, 6 unresolved discussions (waiting on @byrnedj, @igchor, and @vinser52)
> 
> _[`cachelib/allocator/BackgroundMover.h` line 20 at r4](https://reviewable.io/reviews/intel/cachelib/20#-NDe6Z4w3QkpUs7XjTX9:-NDe6Z4w3QkpUs7XjTXA:b-75ywm7) ([raw file](https://github.com/intel/cachelib/blob/b31c5adcb19a9d315a4a818a8238f6eec0c1d3c8/cachelib/allocator/BackgroundMover.h#L20)):_
> 
> > ```objc
> > #include <folly/concurrency/UnboundedQueue.h>
> > #include <gtest/gtest_prod.h>
> > ```
> 
> why is this needed?

It's not - it was leftover from development. I have removed it.

> 
> _[`cachelib/allocator/MMTinyLFU-inl.h` line 235 at r4](https://reviewable.io/reviews/intel/cachelib/20#-NDe6yPvDyVSutdrXzkv:-NDe6yPvDyVSutdrXzkw:b-3u8qk4) ([raw file](https://github.com/intel/cachelib/blob/b31c5adcb19a9d315a4a818a8238f6eec0c1d3c8/cachelib/allocator/MMTinyLFU-inl.h#L235)):_
> 
> > ```objc
> > void
> > MMTinyLFU::Container<T, HookPtr>::withPromotionIterator(F&& fun) {
> >   throw std::runtime_error("Not supported");
> > ```
> 
> Can you maybe open some issue to implement this in future?

Sure, done.

> 
> _[`cachelib/external/fbthrift` line 1 at r4](https://reviewable.io/reviews/intel/cachelib/20#-NDe6Fum1zrMHycpmBpA:-NDe6Fum1zrMHycpmBpB:b-r7568g) ([raw file](https://github.com/intel/cachelib/blob/a4c4ab370780615027361cd9ce0c4840162d99db/cachelib/external/fbthrift#L1)):_
> 
> > ```
> > Subproject commit 74f3a8fb00e3963d1e6f2361fce722688d500417
> > ```
> 
> Can you revert updating the dependencies? It will cause conflicts later

Yes, sorry about that. I have reverted them.

igchor reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> there is just one test failing on the CI (not sure if it's random or not)

> *[Reviewable](https://reviewable.io/reviews/intel/cachelib/20#-:-NDi_mI_6HfGVug0oM5V:bazcoar)* status: 22 of 28 files reviewed, 4 unresolved discussions (waiting on @byrnedj, @igchor, and @vinser52)



<!-- Sent from Reviewable.io -->

byrnedj commented:
When I run the tests locally I do not see any failures. 
[test_results.txt](https://github.com/intel/CacheLib/files/9802017/test_results.txt)

vinser52 reviewed (changes requested):
> Reviewed 11 of 24 files at r1, 4 of 10 files at r2, 1 of 3 files at r3, 2 of 8 files at r4, 1 of 2 files at r5, 3 of 3 files at r6.
> *[Reviewable](https://reviewable.io/reviews/intel/cachelib/20)* status: 27 of 28 files reviewed, 2 unresolved discussions (waiting on @byrnedj and @igchor)

___
*[`MultiTierDataMovement.md` line 15 at r1](https://reviewable.io/reviews/intel/cachelib/20#-NCRgZd6BPIY5ML4C33O:-NEaZJlW5H-ebGBj6qkx:b-m0yo2s) ([raw file](https://github.com/intel/cachelib/blob/a47446594fccd7cd7f5c9f341f628dff2fa5e6db/MultiTierDataMovement.md#L15)):*
<details><summary><i>Previously, vinser52 (Sergei Vinogradov) wrote…</i></summary><blockquote>

Do I understand correctly that tiering would not work in that case? Or we are going to use first-free policy for allocations?
</blockquote></details>

Should it be renamed to `disableMovement`. According to the current description, it looks like we are prohibiting item movement between tiers. As I remember in the legacy implementation this option does not allow throwing away any item implicitly by eviction from the cache. Now according to the comment, the item can be removed by eviction.
___
*[`cachelib/allocator/BackgroundMover.h` line 97 at r6](https://reviewable.io/reviews/intel/cachelib/20#-NEa_BkwA8ySftVR1ZFH:-NEa_BkwA8ySftVR1ZFI:bida1m2) ([raw file](https://github.com/intel/cachelib/blob/d16e6065906c233e5fc5a1ca7d7b53425ea06041/cachelib/allocator/BackgroundMover.h#L97)):*
> ```objc
>   AtomicCounter totalBytesMoved{0};
> 
>   std::vector<std::tuple<TierId, PoolId, ClassId>> assignedMemory_;
> ```

Since we are using `std::tuple<TierId, PoolId, ClassId>` definition in multiple places, should it be some alias for that, e.g. `using memory_descriptor_type = std::tuple<TierId, PoolId, ClassId>;`. In case we want to change this type in the future it is only a single place where it is defined.
Probably, it is even better to have a separate class/struct instead of tuple, but it is up to you.


<!-- Sent from Reviewable.io -->

byrnedj reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/intel/cachelib/20)* status: 20 of 28 files reviewed, 2 unresolved discussions (waiting on @igchor and @vinser52)

___
*[`MultiTierDataMovement.md` line 15 at r1](https://reviewable.io/reviews/intel/cachelib/20#-NCRgZd6BPIY5ML4C33O:-NEr2Rqm4pbl9kkP90gM:bbnzu8s) ([raw file](https://github.com/intel/cachelib/blob/a47446594fccd7cd7f5c9f341f628dff2fa5e6db/MultiTierDataMovement.md#L15)):*
<details><summary><i>Previously, vinser52 (Sergei Vinogradov) wrote…</i></summary><blockquote>

Should it be renamed to `disableMovement`. According to the current description, it looks like we are prohibiting item movement between tiers. As I remember in the legacy implementation this option does not allow throwing away any item implicitly by eviction from the cache. Now according to the comment, the item can be removed by eviction.
</blockquote></details>

It appears that disableEviction will be deprecated. I have removed this from the documentation. 
___
*[`cachelib/allocator/BackgroundMover.h` line 97 at r6](https://reviewable.io/reviews/intel/cachelib/20#-NEa_BkwA8ySftVR1ZFH:-NEr34AXBRQGMbsgF4VZ:b-896fix) ([raw file](https://github.com/intel/cachelib/blob/d16e6065906c233e5fc5a1ca7d7b53425ea06041/cachelib/allocator/BackgroundMover.h#L97)):*
<details><summary><i>Previously, vinser52 (Sergei Vinogradov) wrote…</i></summary><blockquote>

Since we are using `std::tuple<TierId, PoolId, ClassId>` definition in multiple places, should it be some alias for that, e.g. `using memory_descriptor_type = std::tuple<TierId, PoolId, ClassId>;`. In case we want to change this type in the future it is only a single place where it is defined.
Probably, it is even better to have a separate class/struct instead of tuple, but it is up to you.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

igchor reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 8 of 8 files at r7.
> *[Reviewable](https://reviewable.io/reviews/intel/cachelib/20#-:-NEvLXcRDSGggYYUhWT6:bnfp4nl)* status: all files reviewed (commit messages unreviewed), 2 unresolved discussions (waiting on @vinser52)



<!-- Sent from Reviewable.io -->
